src/Homotopy/Truncation/SetTruncation.ard

```diff
@@ -0,0 +1,37 @@
+\import HLevel
+\import Homotopy.Truncation
+\import Paths
+
+\data SetTrunc (A : \Type)
+  | truncSet A
+  | squashSet (x y : SetTrunc A) (p q : x = y) (i j : I) \elim i, j {
+    | left, j => p @ j
+    | right, j => q @ j
+    | i, left => x
+    | i, right => y
+  }
+  \where {
+    \func test {A : \Type} (B : SetTrunc A -> \Type) (BSet : \Pi (x : SetTrunc A) -> isSet (B x)) =>
+      ofHLevel=>ofHLevel'_-1+ (_, B) 1
+
+    -- | Lemma 6.9.1 in HoTT book
+    \func elim {A : \Type} (B : SetTrunc A -> \Type) (BSet : \Pi (x : SetTrunc A) -> isSet (B x))
+               (g : \Pi (x : A) -> B (truncSet x)) (a : SetTrunc A) : B a \elim a
+      | truncSet x => g x
+      | squashSet x y p q i j =>
+        \let | ind_x => elim B BSet g x
+             | ind_y => elim B BSet g y
+             | map_p => path {\lam k => B (p @ k)} (\lam k => elim B BSet g (p @ k))
+             | map_q => path {\lam k => B (q @ k)} (\lam i => elim B BSet g (q @ i))
+             | p=w=q => path (\lam k => path (squashSet x y p q k))
+        \in ofHLevel=>ofHLevel'_-1+ (_, B) 1 BSet ind_x ind_y p=w=q map_p map_q @ i @ j
+
+    \func elim2 {A : \Type} (B : \Pi (_ _ : SetTrunc A) -> \Type) (BSet : \Pi (x y : SetTrunc A) -> isSet (B x y))
+                (g : \Pi (x y : A) -> B (truncSet x) (truncSet y)) : \Pi (a b : SetTrunc A) -> B a b =>
+      elim (\lam a => \Pi (b : SetTrunc A) -> B a b) (\lam a => HLevels-pi (B a) 1 (BSet a))
```

Comment:
It's not inferred here too @valis

Reply:
Yes, it equals `zero` here and `zero` is a constructor.

Reply:
Isn't it equal to `suc zero` ？

Reply:
But it's not inferred

Reply:
Ah, yes, `suc zero`. And `suc` is a constructor.

Reply:
But since it's not inferred, shouldn't we make it explicit? If you replace 1 with _ it'll show type error

Reply:
Implicit arguments are not always inferable. We should make them implicit if they are inferable most of the time.

Reply:
It's only inferable in the definition itself

Reply:
I mean its recursive call in the definition

Reply:
> It can be inferred either from the type of p or from the result type. I see 3 usages in Homotopy.Truncation where it is inferred. There is only one usage in Equiv.Fiber where it is also constant, so it is not inferred.

Reply:
Oh, seems I've missed some usages

Reply:
Fixed